### ISSUE TITLE
Colin McRae Rally 04: sound bugs while EAX is on

### ISSUE THREAD

**Issue by francescoaracu** [kcat/dsoal#69]
As I said in the title, trying to play this game with EAX on leads to some bugs in sound playing. Notably:
- heavy reverb;
- pacenotes are called at the start and for some time in the stages, and then suddenly stop.

Attaching logs for easing troubleshooting purposes.
[CMR04_logs.zip](https://github.com/kcat/dsoal/files/9530547/CMR04_logs.zip)

**Comment by kcat** [comment]
Try using a newer DSOAL (found in [the Actions tab](https://github.com/kcat/dsoal/actions/runs/2620237704)). That log seems to be using an older version before it started using OpenAL's EAX extensions.

**Comment by francescoaracu** [comment]
Used the newer DSOAL version and also the latest OpenAL Soft (from its repo Actions tab), but it's the same strange behavior...

Attaching logs.
[CMR04_logs_updated.zip](https://github.com/kcat/dsoal/files/9530963/CMR04_logs_updated.zip)

**Comment by kcat** [comment]
A couple things I notice. The game uses deferred buffers, and at some point, runs out of hardware sources, which causes some Play calls to fail since it can't get a source. It seems to ignore the failure and set some EAX properties (which in turn causes an OpenAL error since it tries to set a source property without a valid source).

I'm not sure how much of this is normal behavior for the game, though. Perhaps DSOAL needs to more readily return sources for stopped/paused buffers (that would cause the buffer's EAX state to be lost, but maybe that's proper behavior for non-playing deferred sources).

Are you able to make a recording of the game where the audio problem starts?

**Comment by francescoaracu** [comment]
> Are you able to make a recording of the game where the audio problem starts?
I will try to do it today.

Anyway, could you think that the hardware sources problem could be (at least) mitigated with a modified alsoft.ini? I don't have the file in the game .exe's folder.

**Comment by kcat** [comment]
> Anyway, could you think that the hardware sources problem could be (at least) mitigated with a modified alsoft.ini?

No, there's a fixed limit of 128 "hardware" sources since some games don't like more than that and won't recognize hardware acceleration if there's more.

**Comment by francescoaracu** [comment]
I re-tried to play the game with [this release](https://github.com/ThreeDeeJay/dsoal/releases/tag/0.9.3) of DSOAL and OpenAL Soft and the problems are still there.

Video proof is [here](https://youtu.be/CmXBG8wLMO0); other than missing notes by the copilot, even some sounds (e.g. when crashing) seems missing. Timestamps of the missing sounds are in the description.

**Comment by francescoaracu** [comment]
I was just wondering if trying an older version worked and... it works! I used `r422` from [here](http://vaporeon.io/hosted/dsoal-builds/old/) and all sounds seemed to be here. A little notation: copilot volume seems mixed down, with engine sound on the contrary boosted a bit. Reverb is working correctly set to 0 in `alsoft.ini`.

**Comment by kcat** [comment]
It's probably due to the earlier version having a larger number of emulated hardware buffers (up to 256, instead of 128), so it doesn't run out like in the recent versions. I don't think that's a true fix, it just manages to work around the issue.

I think a more proper fix may be to remove the source from deferred hardware buffers after they finish playing. Or allow "stealing" sources from deferred hardware buffers as-needed.

**Comment by francescoaracu** [comment]
Tried [latest release](https://github.com/ThreeDeeJay/dsoal) from ThreeDeeJay's fork but the problem is still there. Don't have any logs unfortunately.

**Comment by Lili1228** [comment]
Just going to say here that the pacenotes problem exists in Colin 2005 as well,

**Comment by Pipi86** [comment]
> Just going to say here that the pacenotes problem exists in Colin 2005 as well,

Yes (also the missing car suspension and crashing sounds), the only way i've found to workaround this issue is to run the game with the launch parameter NOHARDWARESOUND (works for both colin mcrae rally 04/05. Downside is there's no more 3d sound nor that nice reverb effects anymore, but you can hear everything.
